### ISSUE TITLE
Learn local setup not working as expected

### ISSUE THREAD

**Issue by sanyaade-teachings** [AdaCore/learn#442]
Hi Everybody,

I stacked up a local version of the Learn to for teaching but I am having small issues getting the program to work.

Please see below: Error message below:
Console Output:
**The machine running the examples may not be available or is busy, please try again now or come back later.**

No matter what I do the above error message is shown and no compilation is taking place :(
![Screenshot from 2020-11-15 10-56-36](https://user-images.githubusercontent.com/4993211/99183130-9d3f8a80-2731-11eb-9aac-46cf11ffe6f8.png)

I believe, I am doing something wrong or some setting are wrong between the fronted-end and back-end server

I am running this on Centos 7.5 Linux. Everything is running on twelve cores Dell server machine. Host port changed to 8099 and I could access Learn at: http://localhost:8099/

Why do I need to do this? I needed to customise Learn as I am currently teaching my 14-18 year old students Ada at my College. I need to modify and adapt the interface to meet that age level and to also support my coming Ada teaching classes so I needed to spin up an online Ada coding platform for younger age group. 

Currently they are using a portable NotePad++ with Gnat Compilers but some students have Macbook, so having an online IDE to support them will be a good direction to go. Many thanks!

I am surfing the internet for more alternatives that may be possible as long as I can customise the tools to meet my students level and needs. Anything! Docker based only, Docker-compose, Singularity, etc... Many thanks!

Thank you for your support!

Hear from you soon!

God blesses!!!

Best regards,
Sanyaade

**Comment by gusthoff** [comment]
Hello Sanyaade,

Thanks a lot for your interest in Ada!

Before I can help you, I'd need to know whether you're using the solution based on Vagrant VMs described in the `README.md` file.

**Comment by sanyaade-teachings** [comment]
@gusthoff 

Many thanks for your quick reply.

Yes! I was using Vagrant VMs and follow the steps given in the README.md  

Many thanks for your support!

God blesses!!!

Best regards,
Sanyaade

**Comment by gusthoff** [comment]
Thanks for the information. This error message usually indicates that the backend (which you start with `vagrant ssh server`) is not running. Could you please check which error messages the backend is giving you? You can see that in the shell that is currently running the backend.

**Comment by sanyaade-teachings** [comment]
@gusthoff 

Both the backend and frontend are running: Please see below for the screenshots of my Virtualbox below
![image](https://user-images.githubusercontent.com/4993211/99186857-097ab800-274b-11eb-8f14-f27dc3fe149e.png)

Frontend (web):
![image](https://user-images.githubusercontent.com/4993211/99186890-4c3c9000-274b-11eb-9786-dabbd3a4defb.png)

Backend(Server):
![image](https://user-images.githubusercontent.com/4993211/99186923-85750000-274b-11eb-8321-a598ece8418c.png)

Both are running on the same server hardware.

VirtualBox Network Settings:
Server:
![image](https://user-images.githubusercontent.com/4993211/99187029-18ae3580-274c-11eb-837b-bd0e0e4d9081.png)

Web:
![image](https://user-images.githubusercontent.com/4993211/99187080-632fb200-274c-11eb-8492-e4a2669c76d7.png)

It has been running for over 4 days since being installed.

Is there other installation methods apart from Vagrant VMs? Many thanks!

Hear from you soon!

Many thanks for your support!

God blesses!!!

Best regards,
Sanyaade

**Comment by gusthoff** [comment]
@sanyaade-teachings, yes, seeing the machines running in VirtualBox is a good indicator. However, to get more information about this issue, you'd need to verify the shell / terminal where you SSH'ed into the VM. There, if the backend process is running properly, you should see a heartbeat being displayed every 20 seconds or so. Also, when clicking on the "run" button (in the browser), you should see messages from the backend in the terminal. If something is wrong with the VM, those messages should indicate which error is happening.

> Is there other installation methods apart from Vagrant VMs? 

No, using Vagrant VMs is the only method we currently support.

**Comment by sanyaade-teachings** [comment]
@gusthoff 

Okay! I will try those steps out. Its asking me for the login and password. I have to get the scrap pad where I wrote those.

I will get back to you with my result.

Many thanks for your supporting.

Speak to you soon!

God blesses!!!

Best regards,
Sanyaade

**Comment by Robert-Tice** [comment]
Hi Sanyaade,

It looks like you have changed the default ports which is likely the cause of the issue. Is there a reason you have changed them?

We would recommend changing the ports back to the original values and trying again. There are dependencies in the website build that rely on static port values. Let us know if changing the ports back to default works.

If you can't do this and you need to use different ports, let us know and I can show where else in the build environment you'll need to modify values.

**Comment by sanyaade-teachings** [comment]
@Robert-Tice 
Many thanks for getting in touch. Sorry just coming back here due to long teaching duration online.

Regarding Port: I have to change the port value as those values are already in use for few years and to avoid getting port in use errors.

**Re:** If you can't do this and you need to use different ports, let us know and I can show where else in the build environment you'll need to modify values. **I will be grateful to go along with this option. Many thanks!** 

Please show me where else I need to modify port values.  Much appreciated!

Many thanks for all our support!

Hear from you soon!

God blesses!!!

Best regards,
Sanyaade

**Comment by Robert-Tice** [comment]
First thing: Please do not use the vagrant setups in production. They are not intended for this purpose. They have been designed to use locally only while developing content and functionality for the learn website. There are non-trivial security implications that need to be considered to properly deploy this set up in production.

If you do need to modify the ports you will need to change a few things: 

- The web page dev server by default uses port 8080 in the vm and is mapped to host port 8080. It looks like you have already figured out how to map this to a different host port. You can also change the default port being used in the vm for the webpack dev server by changing the port setting in the [webpack.dev.js file](https://github.com/AdaCore/learn/blob/588caf503c39a182ec30f25d28254bdf541b514e/frontend/webpack.dev.js#L19).
- The api server by default uses port 8000 in the vm and is mapped to port 8000 on the host. The web pages need to know which port the api server is using. This is done using the HOST_PORT variable in the [Makefile](https://github.com/AdaCore/learn/blob/588caf503c39a182ec30f25d28254bdf541b514e/frontend/Makefile#L16). You should set this to be the port that the host is using. So if change the vagrant mapping to map the vm port 8000 to host port 6000, you would change the Makefile variable to 600. That will tell the webpage builder to hardcode that port in the webpages during build. 
- You shouldn't need to change the actual port that the api server is using. The port 8000 is the default port being used by the Flask dev server. If you do need to change this for some reason, you can modify the [dev_server.sh](https://github.com/AdaCore/learn/blob/master/backend/dev_server.sh) script to add your required port to the `flask run` command.

**Comment by sanyaade-teachings** [comment]
@Robert-Tice : 

Many thanks for your brilliant support.
I finally get it working just now following your directions.
What I did:

1.  modify  HOST_PORT variable in the **Makefile** as you have stated above. In my case from 8000 to 8007 ( I think this one did the trick)
2. modify FLASK_RUN_PORT=8007 in **backend/.flaskenv**  file

These are in addition to what I have previously done!  I have already changed in my own case:

- web port from 8080 --> 8099
- webpack.dev.js port from 8080 --> 8099
- backend server port from 8000 --> 8007

Next follow  the instructions on README file to get things running.

Many thanks to you and everyone at AdaCore. Much appreciate!

I will be working now on modding it for my students. I would not mind to break things as long as they can use it for their Ada learning. 

I am also looking at other Online-Compilers including those that can be included in LMS like Moodle, openEdx, Canvas, etc...
As I am teaching younger age group Ada, the environment need to be more interactive and appeal to their genre.

Once again, many thanks!

God blesses!!!

Best regards,
Sanyaade
